[COVERAGE RUN PIPE] Expand start-test(run) command by coverage-mode

mykhailiukVitalii opened the pull request:
New Coverage task  start-test run mode (coverage pipe) V3: 
- separate "coverage" pipe with `prepareRun(params)` function (move coverage to pipe);
- combine filter & coverage to one run section (cleanup `cli.js`);
- fix git command to 2 different type:  `git diff master --name-only` or `git diff <branch-name> --name-only` cmd;
- use `minimatch js` lib to parse `.yml` file.

Short Coverage pipe functionality description:
- Retrieved modified files using `git diff --name-only` cmd;
- Validated coverage file  & parsed relevant coverage data;
- Extracted test IDs and generated a full `--grep` command for targeted test runs;
(original ticket - https://github.com/testomatio/playtika-issues/issues/32 )

Example:
- command: `TESTOMATIO_URL=http://xxx TESTOMATIO=tstxxx npx @testomatio/reporter run "npx codeceptjs run" --filter "coverage:file=coverage/coverage.yml, diff=my-branch"`
- `coverage.yml` file example: 
```yml
// coverage.yml
todomvc-tests/helpers/**:
  - "@S171a8408"
  - "@T091ed49e"
todomvc-tests/edit-todos_test.js:
  - "@T091ed49e"
todomvc-tests/pages/**/*.js:
  - "tag:@step-06"
  - "@Safaa0ab4"

```
Future updates:
- create a new test plan at the end??? - **WIP**
- possible duplicates of tests in case if `--grep "(Safaa0ab4|Tc85d62e4)"` and suite includes this test?? - **WIP**
- deeper testing by user??? - **WIP**

mykhailiukVitalii committed:
[COVERAGE PIPE] Coverage pipe task - small --filter docs fix to correct tag-name case

mykhailiukVitalii committed:
[COVERAGE PIPE] Coverage pipe task - global updates by adding a new coverage pipe to system

mykhailiukVitalii committed:
[COVERAGE PIPE] Coverage pipe task - small log message updates: entries instead of tests

mykhailiukVitalii commented:
Review by Michael options:
- introduce filter syntax:
```
npx @testomatio/reporter run "npx jest" --filter "coverage:file=testomatio-coverage.yml,diff=main"
npx @testomatio/reporter run "npx jest" --filter "coverage:diff=main,file=testomatio-coverage.yml"
npx @testomatio/reporter run "npx jest" --filter "coverage:file=coverage.yml"
```

Behavior Summary table of possible command:

| Option | Git command | Notes |
| --- | --- | --- |
| --filter "coverage:file=coverage/coverage.yml,diff=new-branch" | ✅ git diff new-branch --name-only | - |
| --filter "coverage:file=diff=master,coverage.yml" | ✅ git diff master --name-only | - |
| --filter "coverage:file=coverage/coverage.yml" | ✅ git diff master --name-only | default branch = "master" |
| --filter "coverage:file=coverage.yml,dif=noexist-branch" | ❌ Git command failed ...| - |
| --filter "coverage:file=no-exist-coverage.yml" | ❌ Coverage file not found: <>filename>.yml | - |
| --filter "coverage:filepath=coverage.yml" | 🚫 Missing required parameter: "file"... | - |
| --filter "coverage:diff=my-branch" | 🚫 Missing required parameter: "file"...| - |

Example of `cli` usage:
- in case if valid command from the table above (but without execution)
<img width="1867" height="696" alt="Screenshot-example-pipe-1-1" src="https://github.com/user-attachments/assets/58b97fca-9e00-4924-96b1-0af9bd93857f" />

in case if invalid command (but without execution)
<img width="1862" height="365" alt="Screenshot-example-pipe-1-2" src="https://github.com/user-attachments/assets/e822f618-99e4-4009-8c5b-81e653ffc7db" />

mykhailiukVitalii commented:
Next task short fixes:
- write unit tests for `coverage.js` pipe as for `testomatio.fs - prepareRun()` ?? - **WIP**
- linter fixes ?? - **WIP**
- add examples to `cli.js` docs?? - **WIP**

Question to team:

1. rename `coverage.js` pipe to `changes.js` ???
2. combine `testomatio.prepareRun()` with `coverage.prepareRun` to one format of start command in `cli.js` (like `npx @testomatio/reporter run "npx jest" --filter "coverage:tag-name=smoke"`) ???
@DavertMik

DavertMik reviewed (changes requested):
* Move CLI logic to pipes
* End user documentation in docs/ to use that feature with some examples

mykhailiukVitalii committed:
[COVERAGE PIPE] Coverage pipe task - clean cli.js logic

mykhailiukVitalii committed:
[COVERAGE PIPE] Coverage task - add a new documantation to docs/pipes/coverage.md

mykhailiukVitalii committed:
[COVERAGE PIPE] Hude refactoring: move all unnecessary from cli.js to client/pipe files.

mykhailiukVitalii committed:
[COVERAGE PIPE] cli.js doc file small updates.

mykhailiukVitalii committed:
[COVERAGE PIPE] Coverage pipe unit tests

mykhailiukVitalii committed:
[COVERAGE PIPE] linter fixes

mykhailiukVitalii committed:
[COVERAGE PIPE] Small prepareRun() updates + test fixes

mykhailiukVitalii commented:
Huge Coverage task updates:
1. Clean `cli.js` file: to be more understandable
2.  Doc updates: a new `.md` file for "pipes/coverage.md" + small updates to "cli.md" doc file
3. Coverage pipe unit tests (+ small `testomatio` test updates)
4. Coverage pipe linter fixes
5. need to fix conflicts by latest `2.x` branch - :heavy_check_mark:

mykhailiukVitalii committed:
[COVERAGE PIPE] Coverage pipe - fix incorrect tag-name to tag type

mykhailiukVitalii committed:
[COVERAGE PIPE] small fix: move parsePipeOptions() to pipe_utils.js

mykhailiukVitalii committed:
[COVERAGE PIPE] remove createDir() from simple coverage class unit tests

mykhailiukVitalii committed:
[COVERAGE PIPE] quick coverage.md docs update

mykhailiukVitalii commented:
All needed info about updates you can find in `docs/pipes/coverage.md` with examples

DavertMik commented on docs/pipes/coverage.md:
````diff
@@ -0,0 +1,122 @@
+✅ Sample coverage.yml
+
+```yaml
+todomvc-tests/helpers/**:
````
please change 

```suggestion
src/Auth/**.tsx:
```

DavertMik commented on docs/pipes/coverage.md:
```diff
@@ -0,0 +1,122 @@
+  - "@T091e"
+  - "tag:@smoke"
+
+todomvc-tests/edit-todos_test.js:
```
```suggestion
src/Admin/**.tsx:
```

DavertMik commented on docs/pipes/coverage.md:
```diff
@@ -0,0 +1,122 @@
+todomvc-tests/edit-todos_test.js:
+  - "@T0922"
+
+todomvc-tests/pages/**/*.js:
```
```suggestion
src/Checkout/**.tsx
  - "tag:checkout"
```

DavertMik commented on docs/pipes/coverage.md:
````diff
@@ -0,0 +1,122 @@
+
+* Compare changes to a specific Git branch
+```bash
+npx @testomatio/reporter run "npx jest" --filter "coverage:file=coverage/coverage.yml,diff=develop"
````
Add dry run mode to see which tests will be executed without executing them

One of proposed APIs:

```
npx @testomatio/reporter dry-run --filter "coverage:file=coverage/coverage.yml,diff=develop"
```

```
npx @testomatio/reporter run --filter "coverage:file=coverage/coverage.yml,diff=develop"
```

```
npx @testomatio/reporter nothing --filter "coverage:file=coverage/coverage.yml,diff=develop"
```

It should print list of tests

And fix conflicts

mykhailiukVitalii committed:
[COVERAGE PIPE] MR review quick updates: coverage.md & client.js

mykhailiukVitalii committed:
[COVERAGE PIPE] A new option to get only list of matched tests: --filter-list

mykhailiukVitalii committed:
[COVERAGE PIPE] Docs updates for the new --filter-list opt

mykhailiukVitalii commented:
I added a new filter option: get list of matched tests without running execution
```
npx @testomatio/reporter run "npx jest" --filter-list "testomatio:tag-name=frontend"
npx @testomatio/reporter run "npx jest" --filter "testomatio:tag-name=frontend"
or
npx @testomatio/reporter run "npx codeceptjs run" --filter-list "coverage:file=coverage/coverage.yml,diff=master"
npx @testomatio/reporter run "npx codeceptjs run" --filter "coverage:file=coverage/coverage.yml,diff=master"
```

Example:
<img width="1843" height="418" alt="Screenshot-example-pipe-2-2" src="https://github.com/user-attachments/assets/52aa1041-2ff9-4fa0-b7c1-7a2a5817c9f1" />
<img width="1824" height="432" alt="Screenshot-example-pipe-2-1" src="https://github.com/user-attachments/assets/e30f2754-6f29-42ce-9143-22d0889c3e6c" />

DavertMik commented on src/client.js:
```diff
@@ -65,34 +66,50 @@ class Client {
     const { pipe, pipeOptions } = params;
+    const parsedOptions = parsePipeOptions(pipeOptions);
+
+    // Handle Coverage pipe-specific setup
```
this should be in pipe!

mykhailiukVitalii replied:
fixed

DavertMik commented on src/client.js:
```diff
@@ -65,34 +66,50 @@ class Client {
+      process.env.COVERAGE_BRANCH = parsedOptions?.diff || "master";
+    }
+
+    this.pipes = await pipesFactory(params || this.paramsForPipesFactory || {}, this.pipeStore);
```
use params

mykhailiukVitalii replied:
fixed

DavertMik commented on src/bin/cli.js:
```diff
@@ -93,17 +94,51 @@ program
 
+      const SUPPORTED_PIPES = ['coverage', 'testomatio'];
+
+      if (!SUPPORTED_PIPES.includes(pipe)) {
```
we can remove this

mykhailiukVitalii replied:
removed

DavertMik commented on src/bin/cli.js:
```diff
@@ -93,17 +94,51 @@ program
+      const [pipe, ...optsArray] = opts?.filter ? opts?.filter.split(':') : opts?.filterList.split(':');
       const pipeOptions = optsArray.join(':');
 
+      const SUPPORTED_PIPES = ['coverage', 'testomatio'];
```
if !pipe throw Error

mykhailiukVitalii replied:
add message to client.js

DavertMik commented on src/bin/cli.js:
```diff
@@ -93,17 +94,51 @@ program
+          '   • "coverage:<options>" (e.g., --filter "coverage:file=coverage.yml")\n' +
+          '   • "testomatio:<options>" (e.g., --filter-list "testomatio:tag-name=smoke")\n' +
+          '   • "testomatio:<options>" (e.g., --filter "testomatio:tag-name=smoke")\n\n' +
+          '👉 Please refer to the documentation for supported options and usage examples.\n'
```
Check correctness of pipe params inside pipe

mykhailiukVitalii replied:
fixed

DavertMik reviewed (changes requested):
`--filter-list` rename to `--dry-run`

```
npx @testomatio/reporter run 'npx jest' --dry-run 'coverage:....'
```

If `--dry-run` empty print the command that will be executed (with info like pipes, variables, etc). 

```
Command: npx jest
```

If `--dry-run` has option, pass it to pipe
Print additional info like comman dexecuted

```
Command npx jest --grep  .......
```

mykhailiukVitalii committed:
[COVERAGE PIPE] Review fixes: move all code to specific pipe

mykhailiukVitalii committed:
[COVERAGE PIPE] Unit test small fixe after coverage pipe updates

mykhailiukVitalii committed:
[COVERAGE PIPE] filter option pipe warning message in client.js

mykhailiukVitalii commented:
Recent updates:

- cleaned up "cli.js" and "client.js" code by moving all unnecessary logic to the pipe
- fixed "coverage_pipe" tests
- logmessage in case the pipe is not properly set ("client.js")
- coverage pipe is based on class parameters instead of environment variables.

Examples:
<img width="1853" height="466" alt="Screenshot-coverage-7-1" src="https://github.com/user-attachments/assets/0952796d-493c-4e40-93b2-4818dd903c94" />
<img width="1858" height="374" alt="Screenshot-coverage-7-3" src="https://github.com/user-attachments/assets/d6d9d210-245a-4dba-a29b-1fed93425f83" />
<img width="1848" height="394" alt="Screenshot-coverage-7-4" src="https://github.com/user-attachments/assets/3f5585c7-930c-4dbd-990d-205ce71092f7" />

mykhailiukVitalii committed:
[COVERAGE PIPE] Added Full Running Command message to --filter-list option

mykhailiukVitalii commented:
Latest updates:
- Added `Full Running Command: ${command}` message to --filter-list option as extra line after "Matched test/suite IDs:" one
@DavertMik FYI
